File: java/src/test/java/io/github/securekit/RedisRateLimiterStoreTest.java
package io.github.securekit;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import redis.clients.jedis.UnifiedJedis;
import redis.clients.jedis.exceptions.JedisConnectionException;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Requires a real Redis instance reachable at localhost:6379. Tests are
 * skipped automatically if Redis isn't reachable, so the rest of the suite
 * still passes without Redis, but running them for real (not mocked) is
 * what actually proves this store's atomicity claims hold.
 */
class RedisRateLimiterStoreTest {

    private static UnifiedJedis client;

    @BeforeAll
    static void connect() {
        try {
            client = new UnifiedJedis("redis://localhost:6379");
            client.ping();
        } catch (JedisConnectionException e) {
            client = null;
        }
    }

    @AfterAll
    static void disconnect() {
        if (client != null) {
            client.close();
        }
    }

    private String prefix(String testName) {
        return "securekit-test:" + testName + ":";
    }

    @Test
    void allowsUnderLimitAndBlocksOver() {
        Assumptions.assumeTrue(client != null, "no Redis reachable at localhost:6379");
        String prefix = prefix("allowsUnderLimitAndBlocksOver");
        RedisRateLimiterStore store = new RedisRateLimiterStore(client, prefix);
        RateLimiter rl = new RateLimiter(store, 3, 60_000L);
        try {
            assertTrue(rl.allow("client-1"));
            assertTrue(rl.allow("client-1"));
            assertTrue(rl.allow("client-1"));
            assertFalse(rl.allow("client-1"));
        } finally {
            client.del(prefix + "client-1");
        }
    }

    @Test
    void tracksKeysIndependently() {
        Assumptions.assumeTrue(client != null, "no Redis reachable at localhost:6379");
        String prefix = prefix("tracksKeysIndependently");
        RedisRateLimiterStore store = new RedisRateLimiterStore(client, prefix);
        RateLimiter rl = new RateLimiter(store, 1, 60_000L);
        try {
            assertTrue(rl.allow("client-A"));
            assertTrue(rl.allow("client-B"));
        } finally {
            client.del(prefix + "client-A", prefix + "client-B");
        }
    }

    @Test
    void expiresAndResets() throws InterruptedException {
        Assumptions.assumeTrue(client != null, "no Redis reachable at localhost:6379");
        String prefix = prefix("expiresAndResets");
        RedisRateLimiterStore store = new RedisRateLimiterStore(client, prefix);
        RateLimiter rl = new RateLimiter(store, 1, 300L);
        try {
            assertTrue(rl.allow("client-1"));
            assertFalse(rl.allow("client-1"));
            Thread.sleep(400);
            assertTrue(rl.allow("client-1"));
        } finally {
            client.del(prefix + "client-1");
        }
    }

    @Test
    void setsExpiryOnFirstIncrementOnly() throws InterruptedException {
        Assumptions.assumeTrue(client != null, "no Redis reachable at localhost:6379");
        String prefix = prefix("setsExpiryOnFirstIncrementOnly");
        RedisRateLimiterStore store = new RedisRateLimiterStore(client, prefix);
        String redisKey = prefix + "client-1";
        try {
            store.increment("client-1", 3_600_000L);
            long ttl1 = client.pttl(redisKey);

            Thread.sleep(50);
            store.increment("client-1", 3_600_000L);
            long ttl2 = client.pttl(redisKey);

            assertTrue(ttl2 <= ttl1, "TTL should only count down, not reset");
        } finally {
            client.del(redisKey);
        }
    }
}

File: php/tests/RedisRateLimiterStoreTest.php
<?php

declare(strict_types=1);

namespace SecureKit\Tests;

use PHPUnit\Framework\TestCase;
use Predis\Client;
use SecureKit\RateLimiter;
use SecureKit\RedisRateLimiterStore;

/**
 * These tests require a real Redis instance reachable at localhost:6379.
 * They're skipped automatically if Redis isn't reachable, so the rest of
 * the suite still passes without Redis, but running them for real (not
 * mocked) is what actually proves this store's atomicity claims hold.
 */
final class RedisRateLimiterStoreTest extends TestCase
{
    private static function client(): Client
    {
        $client = new Client(['host' => 'localhost', 'port' => 6379]);
        try {
            $client->connect();
        } catch (\Throwable $e) {
            self::markTestSkipped('no Redis reachable at localhost:6379: ' . $e->getMessage());
        }
        return $client;
    }

    public function testAllowsUnderLimitAndBlocksOver(): void
    {
        $client = self::client();
        $prefix = 'securekit-test:' . __FUNCTION__ . ':';
        $store = new RedisRateLimiterStore($client, $prefix);
        $rl = new RateLimiter($store, 3, 60);

        try {
            $this->assertTrue($rl->allow('client-1'));
            $this->assertTrue($rl->allow('client-1'));
            $this->assertTrue($rl->allow('client-1'));
            $this->assertFalse($rl->allow('client-1'));
        } finally {
            $client->del([$prefix . 'client-1']);
        }
    }

    public function testTracksKeysIndependently(): void
    {
        $client = self::client();
        $prefix = 'securekit-test:' . __FUNCTION__ . ':';
        $store = new RedisRateLimiterStore($client, $prefix);
        $rl = new RateLimiter($store, 1, 60);

        try {
            $this->assertTrue($rl->allow('client-A'));
            $this->assertTrue($rl->allow('client-B'));
        } finally {
            $client->del([$prefix . 'client-A', $prefix . 'client-B']);
        }
    }

    public function testExpiresAndResets(): void
    {
        $client = self::client();
        $prefix = 'securekit-test:' . __FUNCTION__ . ':';
        $store = new RedisRateLimiterStore($client, $prefix);
        $rl = new RateLimiter($store, 1, 1); // 1-second window

        try {
            $this->assertTrue($rl->allow('client-1'));
            $this->assertFalse($rl->allow('client-1'));
            usleep(1_100_000);
            $this->assertTrue($rl->allow('client-1'));
        } finally {
            $client->del([$prefix . 'client-1']);
        }
    }

    public function testSetsExpiryOnFirstIncrementOnly(): void
    {
        $client = self::client();
        $prefix = 'securekit-test:' . __FUNCTION__ . ':';
        $store = new RedisRateLimiterStore($client, $prefix);
        $redisKey = $prefix . 'client-1';

        try {
            $store->increment('client-1', 3600);
            $ttl1 = $client->pttl($redisKey);

            usleep(50_000);
            $store->increment('client-1', 3600);
            $ttl2 = $client->pttl($redisKey);

            $this->assertLessThanOrEqual($ttl1, $ttl2, 'TTL should only count down, not reset');
        } finally {
            $client->del([$redisKey]);
        }
    }
}
